### PR TITLE
Issue 490: Add tagsoup support

### PR DESCRIPTION
**Pull request by jsight**
https://github.com/apache/camel-quarkus/issues/490

### CODE REVIEW

**Comment by jsight** [comment]
I don't really understand the failures here. The parent paths appeared to be correct to me, unless I'm misreading something?

File: extensions/tagsoup/deployment/pom.xml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.camel.quarkus</groupId>
+        <artifactId>camel-quarkus-tagsoup-parent</artifactId>
+        <version>0.3.2-SNAPSHOT</version>
```
Comment (lburgazzoli) [inline extensions/tagsoup/deployment/pom.xml]: Should be 0.4.1-SNAPSHOT

File: extensions/tagsoup/pom.xml [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.camel.quarkus</groupId>
+        <artifactId>camel-quarkus-build-parent</artifactId>
+        <version>0.3.2-SNAPSHOT</version>
```
Comment (lburgazzoli) [inline extensions/tagsoup/pom.xml]: Should be 0.4.1-SNAPSHOT

File: extensions/tagsoup/runtime/pom.xml [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.camel.quarkus</groupId>
+        <artifactId>camel-quarkus-tagsoup-parent</artifactId>
+        <version>0.3.2-SNAPSHOT</version>
```
Comment (lburgazzoli) [inline extensions/tagsoup/runtime/pom.xml]: Should be 0.4.1-SNAPSHOT

File: integration-tests/tagsoup/pom.xml [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+    <parent>
+        <groupId>org.apache.camel.quarkus</groupId>
+        <artifactId>camel-quarkus-integration-tests</artifactId>
+        <version>0.3.2-SNAPSHOT</version>
```
Comment (lburgazzoli) [inline integration-tests/tagsoup/pom.xml]: Should be 0.4.1-SNAPSHOT

File: extensions/tagsoup/deployment/pom.xml (hunk 1 above)
Comment (jsight) [reply extensions/tagsoup/deployment/pom.xml]: Oh... thanks :)

**Comment by jsight** [comment]
@lburgazzoli - I think that I've fixed them now.

**Comment by davsclaus** [comment]
ok to test

**Comment by jsight** [comment]
Is this because it needs to be tested in native mode as well?

**Comment by davsclaus** [comment]
Yes when we write that comment about its okay to test, then it does a full test incl native mode too.

File: extensions/tagsoup/runtime/pom.xml [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        <artifactId>camel-quarkus-tagsoup-parent</artifactId>
+        <version>0.4.1-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+
+    <artifactId>camel-quarkus-tagsoup</artifactId>
+    <name>Camel Quarkus :: TagSoup :: Runtime</name>
+
+    <properties>
+        <firstVersion>0.3.0</firstVersion>
```
Comment (ppalaga) [inline extensions/tagsoup/runtime/pom.xml]:
```suggestion
        <firstVersion>0.5.0</firstVersion>
```

File: extensions/tagsoup/runtime/src/main/resources/META-INF/quarkus-extension.yaml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+---
+name: "Camel Quarkus TagSoup"
+description: "Camel TagSoup support"
+metadata:
+  keywords:
+  - "camel"
```
Comment (ppalaga) [inline extensions/tagsoup/runtime/src/main/resources/META-INF/quarkus-extension.yaml]:
```suggestion
  - "camel"
  - "tagsoup"
  - "tidymarkup"
```

File: integration-tests/tagsoup/src/test/resources/test.html [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+-->
+<html>
+<head>
+    <title>Title</title>
+</head>
+<body>
+    <p>Paragraph Contents</p>
```
Comment (ppalaga) [inline integration-tests/tagsoup/src/test/resources/test.html]: I'd say this document is too sane to see TagSoup at doing something really useful. Let's break it a bit  - e.g. let's at least remove the closing tag `</p>` to make it an invalid XML doc.

File: poms/bom/pom.xml [hunk 8]
```diff
@@ -48,10 +48,7 @@
                 <scope>import</scope>
             </dependency>
 
-            <!-- Sort using cd extensions && mvn groovy:execute@sort-poms -->
-            <!-- Dependencies a..z -->
```
Comment (ppalaga) [inline poms/bom/pom.xml]: Please do not remove the above two lines. The second one is important for the `sort-poms` script. 

Also please remove the once removed `<!-- Camel (in alphabetical order) -->` line and re-run `cd extensions && mvn groovy:execute@sort-poms`.

File: integration-tests/tagsoup/src/test/resources/test.html (hunk 7 above)
Comment (jsight) [reply integration-tests/tagsoup/src/test/resources/test.html]: TBH, I wasn't really worried about that. The main thing is that it runs and returns a document. I will add a syntax error, though.

**Comment by jsight** [comment]
I have pushed a fix but this just yields another error. I really do not know how to debug this last one.

File: poms/bom/pom.xml (hunk 8 above)
Comment (jsight) [reply poms/bom/pom.xml]: I hope that you rethink this pattern, but I will try.

**Comment by ppalaga** [comment]
> I have pushed a fix but this just yields another error. I really do not know how to debug this last one.

XPath is untested yet, so seeing it to fail in native mode is rather expected. But it is a good news actually, that the flow fails that late, because it means that TagSoup produces a valid DOM. I think the extension in itself works well, but we need to change the way we test it somehow.

I tried to serialize the DOM using a dummy XSLT but it fails due to issues that need to get fixed in the XSLT extension.

Then I tried to use the DOM API directly to return the content of the text node under /html/body/p and it works. I have sent a PR against your branch: https://github.com/jsight/camel-quarkus/pull/1 Please check whether it makes sense.

**Comment by jsight** [comment]
Quarkus is at 1.0 but doesn't support... xpath? That is an interesting finding. Anyway, I have merged your PR. Hopefully that is all that needs to be done here now.

File: poms/bom/pom.xml (hunk 8 above)
Comment (ppalaga) [reply poms/bom/pom.xml]: Sorry for the inconvenience, but these POMs will gradually contain hundreds of dependencies so sorting them programmatically to be able to figure out what is in there is a must.

**Comment by ppalaga** [comment]
> all that needs to be done here now

Could you please squash all your (and mine) commits in this PR?

**Comment by jsight** [comment]
Done

**Comment by ppalaga** [comment]
ok to test

**Comment by lburgazzoli** [comment]
ok to test

**Pull request merged**